WhateverGreen/kern_rad.cpp

```diff
@@ -183,7 +190,142 @@ void RAD::processKernel(KernelPatcher &patcher, DeviceInfo *info) {
+	if (matching == nullptr) {
+		DBGLOG("igfx", "isRadeonX6000WiredToInternalDisplay null AppleBacklightDisplay");
+		return;
+	}
+	
+	OSIterator *iter = IOService::getMatchingServices(matching);
+	if (iter == nullptr) {
+		DBGLOG("igfx", "isRadeonX6000WiredToInternalDisplay null matching");
+		matching->release();
+		return;
+	}
+	
+	IORegistryEntry* display = OSDynamicCast(IORegistryEntry, iter->getNextObject());
+	if (display == nullptr) {
+		DBGLOG("igfx", "isRadeonX6000WiredToInternalDisplay null display");
+		iter->release();
+		matching->release();
+		return;
+	}
+	
+	OSDictionary* iodispparm = OSDynamicCast(OSDictionary, display->getProperty("IODisplayParameters"));
+	if (iodispparm) {
+		DBGLOG("igfx", "isRadeonX6000WiredToInternalDisplay null IODisplayParameters");
+		iter->release();
+		matching->release();
+		return;
+	}
+	
+	OSDictionary* linearbri = OSDynamicCast(OSDictionary, iodispparm->getObject("linear-brightness"));
+	if (linearbri == nullptr) {
+		DBGLOG("igfx", "isRadeonX6000WiredToInternalDisplay null linear-brightness");
+		iter->release();
+		matching->release();
+		return;
+	}
+	
+	OSNumber* maxbri = OSDynamicCast(OSNumber, linearbri->getObject("max"));
+	if (maxbri == nullptr) {
+		DBGLOG("igfx", "isRadeonX6000WiredToInternalDisplay null max");
+		iter->release();
+		matching->release();
+		return;
+	}
+
+	callbackRAD->maxPwmBacklightLvl = maxbri->unsigned32BitValue();
+	DBGLOG("igfx", "updatePwmMaxBrightnessFromInternalDisplay get max brightness: 0x%x", callbackRAD->maxPwmBacklightLvl);
+
+	iter->release();
+	matching->release();
+}
+
+uint32_t RAD::wrapDcePanelCntlHwInit(void *panel_cntl) {
+	callbackRAD->panelCntlPtr = panel_cntl;
+	callbackRAD->updatePwmMaxBrightnessFromInternalDisplay(); // read max brightness value from IOReg
+	uint32_t ret = FunctionCast(wrapDcePanelCntlHwInit, callbackRAD->orgDcePanelCntlHwInit)(panel_cntl);
+	return ret;
+}
+
+IOReturn RAD::wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute(IOService *framebuffer, IOIndex connectIndex, IOSelect attribute, uintptr_t value) {
+	IOReturn ret = FunctionCast(wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute, callbackRAD->orgAMDRadeonX6000AmdRadeonFramebufferSetAttribute)(framebuffer, connectIndex, attribute, value);
+	if (attribute != (UInt32)'bklt') {
+		return ret;
+	}
+	
+	if (callbackRAD->maxPwmBacklightLvl == 0) {
+		DBGLOG("igfx", "wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute zero maxPwmBacklightLvl");
+		return 0;
+	}
+	
+	if (callbackRAD->panelCntlPtr == nullptr) {
+		DBGLOG("igfx", "wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute null panel cntl");
+		return 0;
+	}
+	
+	if (callbackRAD->orgDceDriverSetBacklight == nullptr) {
+		DBGLOG("igfx", "wrapAMDRadeonX6000AmdRadeonFramebufferSetAttribute null orgDcLinkSetBacklightLevel");
+		return 0;
+	}
+	
+	// set the backlight of AMD navi10 driver
+	callbackRAD->curPwmBacklightLvl = (int)value;
+	int btlper = callbackRAD->curPwmBacklightLvl*100.0 / callbackRAD->maxPwmBacklightLvl;
```

Comment:
No need for floats here. Please remove `.0` part.

Also, `curPwmBacklightLvl` and `maxPwmBacklightLvl` are always unsigned I believe. Change them to `uint32_t` and update the types here, to cast to `(uint32_t)`.